probe_src/probe_py/cli.py

```diff
@@ -116,4 +116,101 @@ def dump(
                     print(op.data)
                 print()
 
+@app.command(
+    context_settings=dict(
+        ignore_unknown_options=True,
+    ),
+)
+def ssh(
+        ssh_args: list[str],
+        debug: bool = typer.Option(default=False, help="Run verbose & debug build of libprobe"),
+):
+    """
+    Wrap SSH and record provenance of the remote command.
+    """
+
+    ssh_cmd = ['ssh']
+
+    if "-i" in ssh_args:
+        rsa_index = ssh_args.index("-i") + 1
+        rsa = ssh_args[rsa_index]
+        ssh_cmd = ssh_cmd[:1] + ['-i',rsa] + ssh_cmd[1:]
+
+    if "-p" in ssh_args:
+        port_index = ssh_args.index("-p") + 1
+        port = ["-p",ssh_args[port_index]]
+        ssh_cmd.extend(port)
+    else:
+        ssh_cmd.extend(["-p",2222])
+
+    if "-h" not in ssh_args:
+        typer.secho("Remote host not defined", fg=typer.colors.RED)
+        raise typer.Abort()
+
+    host_index = ssh_args.index("-h") + 1
+    remote_host = [ssh_args[host_index]] 
+    ssh_cmd.extend(remote_host)
+
+    ssh_cmd_args = ssh_args[host_index + 1:]
+
+    libprobe = project_root / "libprobe/build" / ("libprobe-dbg.so" if debug else "libprobe.so")
+    if not libprobe.exists():
+        typer.secho(f"Libprobe not found at {libprobe}", fg=typer.colors.RED)
+        raise typer.Abort()
+
+    # Create a temporary directory on the local machine
+    local_temp_dir = pathlib.Path(tempfile.mkdtemp(prefix=f"probe_log_{os.getpid()}"))
+
+    # Check if remote platform matches local platform
+    remote_gcc_machine_cmd = ssh_cmd + ["gcc", "-dumpmachine"]
+    local_gcc_machine_cmd = ["gcc", "-dumpmachine"]
+
+    remote_gcc_machine = subprocess.check_output(remote_gcc_machine_cmd).decode().strip()
+    local_gcc_machine = subprocess.check_output(local_gcc_machine_cmd).decode().strip()
+
+    if remote_gcc_machine != local_gcc_machine:
+        raise NotImplementedError("Remote platform is different from local platform")
+
+    # Upload libprobe.so to the remote temporary directory
+    remote_temp_dir_cmd = ssh_cmd + ["mktemp", "-d", "/tmp/probe_log_XXXXXX"]
+    remote_temp_dir = subprocess.check_output(remote_temp_dir_cmd).decode().strip()
+    remote_probe_dir = f"{remote_temp_dir}/probe_dir"
+
+    ssh_g = subprocess.run(ssh_cmd + ['-G'],stdout=subprocess.PIPE)
+    ssh_g_op = ssh_g.stdout.decode().strip().splitlines()
+
+    ssh_pair = []
+    for pair in ssh_g_op:
+        ssh_pair.append(pair.split())
+
+    scp_args = []
+    for pair in ssh_pair:
+        scp_args.extend(['-o',f'{pair[0]}={pair[1]}'])
+
+    scp_cmd = ['scp'] + scp_args + [str(libprobe),f"{remote_host[0]}:{remote_temp_dir}/"]
+    subprocess.run(scp_cmd,check=True)
+
+    # Prepare the remote command with LD_PRELOAD and __PROBE_DIR
+    ld_preload = f"{remote_temp_dir}/{libprobe.name}"
+
+    env = ["env", f"__LD_PRELOAD={ld_preload}", f"__PROBE_DIR={remote_probe_dir}"]
+    proc = subprocess.run(env + ssh_cmd + ssh_cmd_args)
```

Comment:
This should be `ssh_cmd + env + remote_cmd`. The env has to go after the SSH, so the environment variables get set on the remote, not on the local host.